strawberryfields/program.py

```diff
@@ -0,0 +1,897 @@
+        # NOTE: refs have already been through _test_regrefs() in append() and thus should be valid
+        for r in refs:
+            # mark the RegRef as deleted
+            r.active = False
+            #self.reg_refs[r.ind].active = False
+        # NOTE: deleted indices are *not* removed from self.unused_indices
+
+    def lock(self):
+        """Finalize the program.
+
+        When a Program is locked, no more Commands can be appended to it.
+        The locking happens when the program is run, or a successor Program is constructed,
+        in order to ensure that the RegRef state of the Program does not change anymore.
+        """
+        self.locked = True
+
+    def can_follow(self, prev):
+        """Checks if this program can follow the given program.
+
+        This requires that the final RegRef state of the first program matches
+        the initial RegRef state of the second program, i.e. they have the same number
+        number of RegRefs, all with identical indices and activity states.
+
+        Args:
+            prev (Program): preceding program fragment
+        Returns:
+            bool: True if the Program can follow prev
+        """
+        # TODO NOTE unused_indices is not compared here, in order to allow program fragment repetition
+        return self.init_reg_refs == prev.reg_refs
+
+    def _index_to_regref(self, ind):
+        """Try to find a RegRef corresponding to a given subsystem index.
+
+        Args:
+            ind (int): subsystem index
+        Returns:
+            RegRef: corresponding register reference
+        Raises:
+            .RegRefError: if the subsystem cannot be found, or is invalid
+        """
+        # index must be found in the dict
+        if ind not in self.reg_refs:
+            raise RegRefError('Subsystem {} does not exist.'.format(ind))
+        return self.reg_refs[ind]
+
+    def _test_regrefs(self, reg):
+        """Make sure reg is a valid selection of subsystems, convert them to RegRefs.
+
+        A register reference is valid if it is properly recorded in self.reg_refs
+        and has not been deleted. The selection is valid if it contains only
+        valid RegRefs and no subsystem is repeated.
+
+        Args:
+            reg (Iterable[int, RegRef]): subsystem references
+        Returns:
+            list[RegRef]: converted subsystem references
+        Raises:
+            .RegRefError: if an invalid subsystem reference is found
+        """
+        temp = []
+        for rr in reg:
+            # must be either an integer or a RegRef
+            if isinstance(rr, RegRef):
+                # regref must be found in the dict values (the RegRefs are compared using __eq__, which, since we do not define it, defaults to "is")
+                if rr not in self.reg_refs.values():
+                    raise RegRefError('Unknown RegRef.')
+                if self.reg_refs[rr.ind] is not rr:
+                    raise RegRefError('Should never happen!')
```

Comment:
Should we make the error message more clear? What exactly should never happen?

Reply:
Yeah, we should. I'll fix it.